### PR TITLE
Alpaca doc

### PR DESCRIPTION
**Pull request by dlwh**

### CODE REVIEW

File: docs/Replicating-Alpaca.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+### Getting Llama 2
+
+If you haven't already, go to [Llama 2's Hugging Face page](https://huggingface.co/meta-llama/Llama-2-7b-hf) and request access to the model.
+
+Once you have access, go to [Hugging Face's Tokens page](https://huggingface.co/settings/tokens) to get an API token.
+
+## The Alpaca script
+
+We have a [Levanter version](https://github.com/stanford-crfm/levanter/blob/main/examples/alpaca.py) of the [original Alpaca script](https://github.com/tatsu-lab/stanford_alpaca/blob/main/train.py)
```
Comment (Ivan-Zhou) [inline docs/Replicating-Alpaca.md:43]:
```suggestion
We have a [Levanter version](../examples/alpaca.py) of the [original Alpaca script](https://github.com/tatsu-lab/stanford_alpaca/blob/main/train.py)
```

Doesn't need to be in this way, but I feel relative link is cooler :)
I have tested locally to make sure the link works.

File: docs/Replicating-Alpaca.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+## Launching the job
+
+Now we can launch the job. We need just a tiny bit of ceremony to get the Hugging Face and WANDB API tokens in the environment:
+(If you're using Llama 1, you don't need the `HUGGING_FACE_HUB_TOKEN` line.)
+
+```bash
+gcloud compute tpus tpu-vm ssh llama-32 -z us-east1-d --worker=all \
+--command="WANDB_API_KEY=${YOUR TOKEN HERE} \
+HUGGING_FACE_HUB_TOKEN=${YOUR TOKEN HERE} \
+bash levanter/infra/run.sh python \
```
Comment (Ivan-Zhou) [inline docs/Replicating-Alpaca.md:79]: Don't you need `levanter/src/levanter/main/train_lm.py` here?

File: src/levanter/compat/torch_serialization.py [hunk 3]
```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
...
+                    return np.array(arr)
+
+                shardings = [None if i != axis_to_shard else "device" for i in range(len(arr.shape))]
+                sharding = NamedSharding(process_mesh, PartitionSpec(*shardings))
+                out = jax.jit(_identity_fn, out_shardings=sharding)(arr)
+                return np.array(out)
+
+        # need to make sure the model is on *this machine* and *this machine's CPU* before saving
+        model = jax.tree_util.tree_map(lambda arr: get_to_cpu(arr), model)
+        # TODO: it's be nice if safetensors supported an iterator or something so we could do the allgather one at a time
```
Comment (Ivan-Zhou) [inline src/levanter/compat/torch_serialization.py:406]:
```suggestion
        # TODO: it would be nice if safetensors supported an iterator or something so we could do the allgather one at a time
```

File: src/levanter/compat/torch_serialization.py [hunk 4]
```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
...
+            elif isinstance(arr, np.ndarray):
+                return arr
+            elif arr.is_fully_addressable:
+                r = np.array(arr)
+                return r
+            else:
+                # unfortunately, jax's allgather seems to replicate to every device rather than every host
+                # which doesn't work for ~7B parameter models on TPU (assuming we also have optimizer state)
+                # this approach limits us to <64B parameters, but that's good enough for now
+                # we're going to do something a bit fancy, where we shard make a (process, device) mesh,
```
Comment (Ivan-Zhou) [inline src/levanter/compat/torch_serialization.py:383]:
```suggestion
                # we're going to do something a bit fancy, where we shard the model into a (process, device) mesh,
```
Am I right?

File: src/levanter/compat/torch_serialization.py [hunk 5]
```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
...
+                # then look for some axis along which we can shard the array, and then we'll do an allgather
+                # via pjit. If we can't find one, we'll just fully replicate since it probably isn't that big.
+                # TODO: ensure that this mesh arranges devices correctly
```
Comment (Ivan-Zhou) [inline src/levanter/compat/torch_serialization.py:386]: How do we do that?

File: src/levanter/compat/torch_serialization.py [hunk 6]
```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
...
+        def get_to_cpu(arr):
+            if not is_jax_array_like(arr):
+                return arr
+            elif isinstance(arr, np.ndarray):
+                return arr
+            elif arr.is_fully_addressable:
+                r = np.array(arr)
+                return r
+            else:
+                # unfortunately, jax's allgather seems to replicate to every device rather than every host
```
Comment (Ivan-Zhou) [inline src/levanter/compat/torch_serialization.py:380]: What's the relationship between device and host? 

From [TPU's system architecture](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu_device), I found that:
> A TPU device contains four TPU chips interconnected by high-speed interconnects. 
> ...
> Each TPU device is connected to a CPU-based host machine (TPU VM). 

Does it mean 1 host contains 1 device, or not necessary?

Comment (dlwh) [reply src/levanter/compat/torch_serialization.py:380]: 8 devices per host (on tpu v3)

File: src/levanter/compat/torch_serialization.py (hunk 5 above)
Comment (dlwh) [reply src/levanter/compat/torch_serialization.py:386]: eh, i dunno. i'm pretty sure this is guaranteed by JAX

File: src/levanter/compat/torch_serialization.py (hunk 4 above)
Comment (dlwh) [reply src/levanter/compat/torch_serialization.py:383]: yeah sorry, thinko

File: docs/Replicating-Alpaca.md (hunk 1 above)
Comment (dlwh) [reply docs/Replicating-Alpaca.md:43]: it doesn't work with the mkdocs though :(
